yew/src/virtual_dom/vlist.rs

```diff
@@ -94,74 +88,347 @@ impl VDiff for VList {
+                .collect();
+
+            // We will try to match the left vnodes with the right ones, and
+            // store them in `matched_rights`: it is formed of vnodes from
+            // right, in the same order than left. `reused_rights` is used to
+            // know which right nodes have been matched and reused.
+            let mut matched_rights: Vec<Option<VNode>> = vec![None; self.children.len()];
+            let mut reused_rights = FixedBitSet::with_capacity(rights.len());
+
+            // Generator of dummy vnodes, used when stealing vnodes from rights.
+            let make_dummy = || {
+                VNode::VText(VText {
+                    text: String::default(),
+                    reference: None,
+                })
+            };
+
+            // Match all the elements that are not new, but may have moved.
+            let mut new_lefts: HashSet<usize> = HashSet::with_capacity(self.children.len());
+            for ((left_idx, left), matched_right) in self
+                .children
+                .iter()
+                .enumerate()
+                .zip(matched_rights.iter_mut())
+            {
+                let left_key = left.key().as_ref().expect("must have a key");
+                if let Some(right_idx) = rights_lookup.remove(&left_key) {
+                    let right = rights
+                        .get_mut(right_idx)
+                        .expect("the index from the map must be valid");
+                    let right = core::mem::replace(right, make_dummy());
+                    *matched_right = Some(right);
+                    reused_rights.put(right_idx);
                 } else {
-                    new_next_sibling.link(next_sibling.clone());
+                    // This left node is new.
+                    new_lefts.insert(left_idx);
+                }
+            }
+            let new_lefts = new_lefts; // remove mutability
+
+            // Reconciliation loop, i.e. apply the least amount of
+            // transformations to rights to make them identical to lefts.
+            let mut matched_rights = matched_rights.into_iter().peekable();
+            let mut last_next_sibling = NodeRef::default();
+            let mut nodes: Vec<NodeRef> = self
+                .children
+                .iter_mut()
+                .map(|left| {
+                    let ancestor = matched_rights.next().unwrap();
+
+                    // Create a new `next_sibling` reference which points to the next `right` or
+                    // the outer list's `next_sibling` if there are no more `rights`.
+                    let new_next_sibling = NodeRef::default();
+                    if let Some(Some(next_right)) = matched_rights.peek() {
+                        new_next_sibling.set(Some(next_right.first_node()));
+                    } else {
+                        new_next_sibling.link(next_sibling.clone());
+                    }
```

Comment:
If the next right is unmatched (but is not the last one), we enter this branch and will place a mid-list left at the end of the list. I think we need to move rights before the reconciliation loop to avoid this issue.